drivers/dma/dma_iproc_pax_v2.c

```diff
@@ -0,0 +1,1058 @@
+	ring->curr.sync_data.signature = PAX_DMA_WRITE_SYNC_SIGNATURE;
+	ring->curr.sync_data.ring = ring->idx;
+	/* pkt id for active dma xfer */
+	ring->curr.sync_data.opaque = 0x0;
+	/* pkt count for active dma xfer */
+	ring->curr.sync_data.total_pkts = 0x0;
+}
+
+static int init_rm(struct dma_iproc_pax_data *pd)
+{
+	int ret = -ETIMEDOUT, timeout = 1000;
+
+	k_mutex_lock(&pd->dma_lock, K_FOREVER);
+	/* Wait for Ring Manager ready */
+	do {
+		LOG_DBG("Waiting for RM HW init\n");
+		if ((sys_read32(RM_COMM_REG(pd, RM_COMM_MAIN_HW_INIT_DONE)) &
+		    RM_COMM_MAIN_HW_INIT_DONE_MASK)) {
+			ret = 0;
+			break;
+		}
+		k_sleep(K_MSEC(1));
+	} while (--timeout);
+	k_mutex_unlock(&pd->dma_lock);
+
+	if (!timeout) {
+		LOG_WRN("RM HW Init timedout!\n");
+	} else {
+		LOG_INF("PAX DMA RM HW Init Done\n");
+	}
+
+	return ret;
+}
+
+static void rm_cfg_start(struct dma_iproc_pax_data *pd)
+{
+	uint32_t val;
+
+	k_mutex_lock(&pd->dma_lock, K_FOREVER);
+
+	/* set config done 0, enable toggle mode */
+	val = sys_read32(RM_COMM_REG(pd, RM_COMM_CONTROL));
+	val &= ~RM_COMM_CONTROL_CONFIG_DONE;
+	sys_write32(val, RM_COMM_REG(pd, RM_COMM_CONTROL));
+
+	val &= ~(RM_COMM_CONTROL_MODE_MASK << RM_COMM_CONTROL_MODE_SHIFT);
+
+#ifdef CONFIG_DMA_IPROC_PAX_DOORBELL_MODE
+	val |= (RM_COMM_CONTROL_MODE_DOORBELL <<
+		RM_COMM_CONTROL_MODE_SHIFT);
+#elif CONFIG_DMA_IPROC_PAX_TOGGLE_MODE
+	val |= (RM_COMM_CONTROL_MODE_ALL_BD_TOGGLE <<
+		RM_COMM_CONTROL_MODE_SHIFT);
+#endif
+	sys_write32(val, RM_COMM_REG(pd, RM_COMM_CONTROL));
+	sys_write32(RM_COMM_MSI_DISABLE_MASK,
+		    RM_COMM_REG(pd, RM_COMM_MSI_DISABLE));
+
+	val = sys_read32(RM_COMM_REG(pd, RM_COMM_AXI_READ_BURST_THRESHOLD));
+	val &= ~(RM_COMM_THRESHOLD_CFG_RD_FIFO_MAX_THRESHOLD_MASK <<
+		 RM_COMM_THRESHOLD_CFG_RD_FIFO_MAX_THRESHOLD_SHIFT);
+	val |= RM_COMM_THRESHOLD_CFG_RD_FIFO_MAX_THRESHOLD_SHIFT_VAL <<
+	       RM_COMM_THRESHOLD_CFG_RD_FIFO_MAX_THRESHOLD_SHIFT;
+	sys_write32(val, RM_COMM_REG(pd, RM_COMM_AXI_READ_BURST_THRESHOLD));
+
+	val = sys_read32(RM_COMM_REG(pd, RM_COMM_FIFO_FULL_THRESHOLD));
+	val &= ~(RM_COMM_PKT_ALIGNMENT_BD_FIFO_FULL_THRESHOLD_MASK <<
+		 RM_COMM_PKT_ALIGNMENT_BD_FIFO_FULL_THRESHOLD_SHIFT);
+	val |= RM_COMM_PKT_ALIGNMENT_BD_FIFO_FULL_THRESHOLD_VAL <<
+	       RM_COMM_PKT_ALIGNMENT_BD_FIFO_FULL_THRESHOLD_SHIFT;
+
+	val &= ~(RM_COMM_BD_FIFO_FULL_THRESHOLD_MASK <<
+		 RM_COMM_BD_FIFO_FULL_THRESHOLD_SHIFT);
+	val |= RM_COMM_BD_FIFO_FULL_THRESHOLD_VAL <<
+	       RM_COMM_BD_FIFO_FULL_THRESHOLD_SHIFT;
+	sys_write32(val, RM_COMM_REG(pd, RM_COMM_FIFO_FULL_THRESHOLD));
+
+	/* Enable Line interrupt */
+	val = sys_read32(RM_COMM_REG(pd, RM_COMM_CONTROL));
+	val |= RM_COMM_CONTROL_LINE_INTR_EN;
+	sys_write32(val, RM_COMM_REG(pd, RM_COMM_CONTROL));
+
+	/* Enable AE_TIMEOUT */
+	sys_write32(RM_COMM_AE_TIMEOUT_VAL,
+		    RM_COMM_REG(pd, RM_COMM_AE_TIMEOUT));
+	val = sys_read32(RM_COMM_REG(pd, RM_COMM_CONTROL));
+	val |= RM_COMM_CONTROL_AE_TIMEOUT_EN;
+	sys_write32(val, RM_COMM_REG(pd, RM_COMM_CONTROL));
```

Comment:
for doorbell mode to work, don't we need enable the AE locking, and set the group ring count to 1?

Reply:
AE locking is needed if packet extension needs to be supported otherwise for doorbell mode its not required.